katdal/categorical.py

```diff
@@ -17,107 +17,93 @@
-
-    evaluates to False, even though the dicts are equal. On top of this there
-    is also the issue of potentially differing pickle protocols.
+    The object is explicitly non-hashable to force comparisons to be by value
+    instead of by identity for ndarrays.
 
     """
 
-    def __init__(self, hashable_value):
-        self.hashable_value = hashable_value
+    def __init__(self, value):
+        self.unwrapped = value
 
     def __repr__(self):
         """Short human-friendly string representation of wrapper object."""
         return "<katdal.%s { %r } at 0x%x>" % \
-               (self.__class__.__name__, self.unwrap(), id(self))
+               (self.__class__.__name__, self.unwrapped, id(self))
 
     def __str__(self):
         """Longer human-friendly string representation of wrapped object."""
-        return str(self.unwrap())
+        return str(self.unwrapped)
 
     def __eq__(self, other):
         """Equality comparison operator."""
-        if not isinstance(other, HashableValueWrapper):
-            other = HashableValueWrapper.wrap(other)
-        return self.hashable_value == other.hashable_value
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        if isinstance(self.unwrapped, np.ndarray) or isinstance(other, np.ndarray):
+            return np.array_equal(self.unwrapped, other)
+        else:
+            return self.unwrapped == other
 
     def __ne__(self, other):
         """Inequality comparison operator."""
         return not self == other
 
     def __lt__(self, other):
         """Less-than comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped < other
 
     def __gt__(self, other):
         """Greather-than comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped > other
 
     def __le__(self, other):
         """Less-than-or-equal comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped <= other
 
     def __ge__(self, other):
         """Greater-than-or-equal comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped >= other
 
-    def __hash__(self):
-        """Return hash value of wrapped object (guaranteed to be hashable)."""
-        return hash(self.hashable_value)
-
-    @staticmethod
-    def wrap(value):
-        """Wrap generic object to become hashable (pickle implementation)."""
-        return HashableValueWrapper(pickle.dumps(value))
-
-    def unwrap(self):
-        """Extract original object from wrapper (pickle implementation)."""
-        return pickle.loads(self.hashable_value)
+    # Force class to be unhashable to allow value comparison via __eq__ if possible
+    __hash__ = None
```

Comment:
What do you think of making `__hash__` return `hash(self.unwrapped)`? I admit I feel a bit odd about making a class that is only sometimes hashable, but it would let the fast path of `unique_in_order` work on a list of wrappers of hashables.